src/contracts/core/AVSDirectory.sol

```diff
@@ -416,12 +485,167 @@ contract AVSDirectory is
         }
     }
 
+    /**
+     * @notice For a single strategy, allocate magnitude to a list of operator sets and then update freeMagnitude
+     * for the strategy after all allocations
+     * @param operator address to allocate magnitude for
+     * @param allocation magnitude adjustment for a single strategy and multiple operator sets
+     * @param effectTimestamp timestamp when the allocation will take effect
+     */
+    function _allocate(address operator, MagnitudeAdjustment calldata allocation, uint32 effectTimestamp) internal {
+        IStrategy strategy = allocation.strategy;
+        OperatorSet[] calldata operatorSets = allocation.operatorSets;
+        uint64 freeAllocatableMagnitude = freeMagnitude[operator][strategy];
+
+        for (uint256 i = 0; i < operatorSets.length; ++i) {
+            // 1. check freeMagnitude available, that is the allocation of stake is backed and not slashable
+            require(
+                freeAllocatableMagnitude >= allocation.magnitudeDiffs[i],
+                "AVSDirectory.queueAllocations: insufficient available free magnitude to allocate"
+            );
+            // 2. rate limiting of 1 pending allocation at a time
+            // read number of checkpoints after current timestamp
+            (uint224 value, uint256 pos) = _magnitudeUpdate[operator][strategy][operatorSets[i].avs][operatorSets[i]
```

Comment:
`currentMagnitude224` or something? is there an issue with making the lib have a uint64 instead?

Reply:
Whatever unit we decide on I agree we should store that in the checkpoint struct to avoid redundant casting